src/common/valgrind_internal.h

```diff
@@ -457,3 +460,40 @@ extern unsigned _On_valgrind;
 #endif
 
 #endif
+
+#define EMIT_API_LOG(lib, func, suffix)					\
+	do {								\
+		char lib_name[LIB_LOG_LEN] = {0};			\
+		char func_name[FUNC_LOG_LEN] = {0};			\
+		size_t lib_len = strlen(lib);				\
+		size_t func_len = strlen(func);				\
+									\
+		if (lib_len + 1 > LIB_LOG_LEN) {			\
+			VALGRIND_EMIT_LOG("Library name is too long");	\
+			break;						\
+		}							\
+		if (func_len + 1 > FUNC_LOG_LEN) {			\
+			VALGRIND_EMIT_LOG("Function name is too long");	\
+			break;						\
+		}							\
+		strcat(lib_name, lib);					\
+		strcat(lib_name, suffix);				\
+		strcat(func_name, func);				\
+		strcat(func_name, suffix);				\
+									\
+		if (strcmp(suffix, ".BEGIN") == 0) {			\
```

Comment:
I think this information should be provided by the caller, because this check may not be optimized out.